src/combinatorics/quiver/covering.lean

```diff
@@ -0,0 +1,303 @@
+/-
+Copyright (c) 2022 Antoine Labelle, Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Labelle, Rémi Bottinelli
+-/
+import combinatorics.quiver.basic
+import combinatorics.quiver.symmetric
+import combinatorics.quiver.cast
+
+/-!
+# Covering
+
+This file defines coverings of quivers as prefunctors that are bijective on the
+so-called stars and costars at each vertex of the domain.
+
+## Main definitions
+
+* `quiver.star u` is the type of all arrows with source `u`;
+* `quiver.costar u` is the type of all arrows with target `u`;
+* `prefunctor.star φ u` is the obvious function `star u → star (φ.obj u)`;
+* `prefunctor.costar φ u` is the obvious function `costar u → costar (φ.obj u)`;
+* `prefunctor.is_covering φ` means that `φ.star u` and `φ.costar u` are bijections for all `u`;
+* `quiver.star_path u` is the type of all paths with source `u`;
+* `prefunctor.star_path u` is the obvious function `star_path u → star_path (φ.obj u)`.
+
+## Main statements
+
+* `prefunctor.path_star_bijective` states that if `φ` is a covering, then `φ.star_path u` is a
+  bijection for al `u`.
+  In other words, any path in the codomain of `φ` lifts uniquely to its domain.
+  
+  ## TODO
+  
+  Clean up the namespaces by renaming `prefunctor` to `quiver.prefunctor`.
+
+## Tags
+
+Cover, covering, quiver, path, lift
+-/
+
+open function quiver
+
+universes u v w
+
+variables {U : Type*} [quiver.{u+1} U]
+          {V : Type*} [quiver.{v+1} V] (φ : U ⥤q V)
+          {W : Type*} [quiver.{w+1} W] (ψ : V ⥤q W)
+
+/-- The quiver.star at a vertex is the collection of arrows whose source is the vertex. -/
+@[reducible] def quiver.star (u : U) := Σ (v : U), (u ⟶ v)
+
+/-- The quiver.costar at a vertex is the collection of arrows whose target is the vertex. -/
+@[reducible] def quiver.costar (u : U) := Σ (v : U), (v ⟶ u)
+
+@[simp] lemma quiver.star_eq_iff {u : U} (F G : quiver.star u) :
+  F = G ↔ ∃ h : F.1 = G.1, F.2.cast rfl h = G.2 :=
+begin
+  split,
+  { rintro ⟨⟩, exact ⟨rfl, rfl⟩, },
+  { induction F, induction G, rintro ⟨h, H⟩, cases h, cases H,
+    simp only [eq_self_iff_true, heq_iff_eq, and_self], }
+end
+
+@[simp] lemma quiver.costar_eq_iff {u : U} (F G : quiver.costar u) :
+  F = G ↔ ∃ h : F.1 = G.1, F.2.cast h rfl = G.2 :=
+begin
+  split,
+  { rintro ⟨⟩, exact ⟨rfl, rfl⟩, },
+  { induction F, induction G, rintro ⟨h, H⟩, cases h, cases H,
+    simp only [eq_self_iff_true, heq_iff_eq, and_self], }
+end
```

Comment:
Those are general lemma about `sigma`, right?

Reply:
Indeed (though I didn't notice it), but I actually can't find such a lemma making use of transport rather than heq.
```
lemma _root_.sigma.ext_rec {α : Type*} {β : α → Type*} (b₁ b₂ : Σ (a : α), β a) :
  b₁ = b₂ ↔ ∃ h : b₁.1 = b₂.1, h.rec_on b₁.2 = b₂.2 :=
begin
  split,
  { rintro ⟨⟩, exact ⟨rfl, rfl⟩, },
  { induction b₁, induction b₂, rintro ⟨h, H⟩, cases h, cases H,
    simp only [eq_self_iff_true, heq_iff_eq, and_self], }
end

@[simp] lemma quiver.star_eq_iff {u : U} (F G : quiver.star u) :
  F = G ↔ ∃ h : F.1 = G.1, F.2.cast rfl h = G.2 :=
begin
  apply sigma.ext_rec,
end
```
This ^ works. What shall I do?

Reply:
Looks good. I think you should add this lemma to `data.sigma.basic` and drop the `quiver.star`/`quiver.costar` versions because they are reducible and you don't want those lemmas to be simp.

Reply:
But then I lose the statement in terms of `quiver.hom.cast`, which is a bit bad API-wise I believe.

Reply:
Also, will slipping `_root_.sigma.ext_rec` in this PR, and making a porting PR once this one is merged work?

Reply:
I don't think that's bad API-wise, given that this is a reducible definition.

Please put it in the right place on the first go. Forward-porting isn't so hard 😉

Reply:
OK, I called the lemma `ext_iff'` because `ext_rec` didn't make much sense imo.